src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -6355,6 +6408,416 @@ register_handler(int sig, struct sigaction *old_handler)
+	struct daos_event *eps[AIO_EQ_DEPTH + 1] = {0};
+	struct aio_ev     *p_aio_ev;
+
+	ev_dequeue_batch(ctx, min_nr, nr, events, num_ev);
+	if (*num_ev >= min_nr)
+		return;
+
+	/* loop over all EQs */
+	for (i = 0; i < aio_eq_count; i++) {
+		if (atomic_load_relaxed(&aio_eq_list[i].n_op_queued) == 0)
+			continue;
+
+		rc = daos_eq_poll(aio_eq_list[i].eq, 0, DAOS_EQ_NOWAIT, AIO_EQ_DEPTH, eps);
+		if (rc < 0)
+			DL_ERROR(rc, "daos_eq_poll() failed");
+
+		for (j = 0; j < rc; j++) {
+			if (eps[j]->ev_error) {
+				DS_ERROR(eps[j]->ev_error, "daos_eq_poll() error");
+			} else {
+				atomic_fetch_add_relaxed(&aio_eq_list[i].n_op_queued, -1);
+				atomic_fetch_add_relaxed(&aio_eq_list[i].n_op_done, 1);
+				p_aio_ev = container_of(eps[j], struct aio_ev, ev);
+				if (p_aio_ev->ctx == ctx) {
+					/* append to event list */
+					D_MUTEX_LOCK(&ctx->lock);
+					events[*num_ev].obj = p_aio_ev->piocb;
+					events[*num_ev].res = p_aio_ev->piocb->u.c.nbytes;
+					daos_event_fini(&p_aio_ev->ev);
+					(*num_ev)++;
+					D_MUTEX_UNLOCK(&ctx->lock);
+					D_FREE(p_aio_ev);
+				} else {
+					/* need to append context's finished queue */
+					ev_enqueue(p_aio_ev->ctx, p_aio_ev);
+				}
+			}
+		}
+	}
+
+	/* try again after polling */
+	ev_dequeue_batch(ctx, min_nr, nr, events, num_ev);
+	if (*num_ev >= min_nr)
+		return;
+}
+
+static int
+io_getevents_sys(io_context_t ctx, long min_nr, long max_nr, struct io_event *events,
+		 struct timespec *timeout)
+{
+	return syscall(__NR_io_getevents, ctx, min_nr, max_nr, events, timeout);
+}
+
+int
+io_getevents(io_context_t ctx, long min_nr, long nr, struct io_event *events,
+	     struct timespec *timeout)
+{
+	aio_ctx_t      *aio_ctx_obj = (aio_ctx_t *)ctx;
+	io_context_t    ctx_real    = aio_ctx_obj->ctx;
+	int             op_done     = 0;
+	struct timespec times_0;
+	struct timespec times_1;
+	struct timespec dt;
+
+	if (next_io_getevents == NULL)
+		next_io_getevents = io_getevents_sys;
+
+	if (!hook_enabled)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+
+	if (aio_ctx_obj->depth == 0)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+	if (!aio_ctx_obj->on_dfs)
+		return next_io_getevents(ctx_real, min_nr, nr, events, timeout);
+	if (!aio_ctx_obj->inited) {
+		DS_ERROR(EINVAL, "event queue is not initialized yet");
+		return (-EINVAL);
+	}
+	if (timeout)
+		clock_gettime(CLOCK_REALTIME, &times_0);
+
+	if (min_nr > AIO_EQ_DEPTH)
+		min_nr = AIO_EQ_DEPTH;
+
+	while (1) {
+		aio_poll_eqs(aio_ctx_obj, min_nr, nr, events, &op_done);
+		if (op_done >= min_nr)
+			return op_done;
+		if (timeout) {
+			clock_gettime(CLOCK_REALTIME, &times_1);
+			dt.tv_sec  = times_1.tv_sec - times_0.tv_sec;
+			dt.tv_nsec = times_1.tv_nsec - times_0.tv_nsec;
+			if (dt.tv_nsec < 0) {
+				dt.tv_sec--;
+				dt.tv_nsec += 1000000000L;
+			}
+			if ((dt.tv_sec > timeout->tv_sec) ||
+			    ((dt.tv_sec == timeout->tv_sec) && ((dt.tv_nsec > timeout->tv_nsec)))) {
+				return op_done;
+			}
```

Comment:
why implement your own timeout mechanism if daos_eq_poll supports it already?

Reply:
Because we need to poll multiple event queues and return once the number of finished requests reaches desired threshold.